zebra-state/src/service.rs

```diff
@@ -679,3 +690,46 @@ impl Service<Request> for StateService {
 pub fn init(config: Config, network: Network) -> BoxService<Request, Response, BoxError> {
     BoxService::new(StateService::new(config, network))
 }
+
+/// Check if zebra is following a legacy chain and return an error if so.
+pub fn legacy_chain_check<I>(
+    tip_height: block::Height,
+    ancestors: I,
+    network: Network,
+) -> Result<(), BoxError>
+where
+    I: Iterator<Item = Arc<Block>>,
+{
+    const MAX_BLOCKS_TO_CHECK: usize = 100;
+
+    // Only do the check if we have an Nu5 activation height.
+    if let Some(nu5_height) = NetworkUpgrade::Nu5.activation_height(network) {
+        // And only do the check if we are above Nu5 activation.
+        if tip_height >= nu5_height {
+            for (count, block) in ancestors.enumerate() {
+                if count >= MAX_BLOCKS_TO_CHECK {
+                    return Err("giving up after checking too many blocks".into());
+                }
+
+                // All transaction `network_upgrade` fields must be consistent with the block height.
+                if block
+                    .check_transaction_network_upgrade_consistency(network)
+                    .is_err()
+                {
+                    return Err("inconsistent network upgrade found in transaction".into());
+                }
```

Comment:
Comment: another way to do this is with `map_err`, but I'm not sure if it's more readable :thinking: 

```suggestion
                block
                    .check_transaction_network_upgrade_consistency(network)
                    .map_err(|_| "inconsistent network upgrade found in transaction")?;
```

Reply:
done at https://github.com/ZcashFoundation/zebra/pull/2366/commits/0587bed76c3d302a01e4a20ed943c2ba11b79214